File: apigateway-service/src/main/java/com/example/apigatewayservice/filter/CorsFilter.java
package com.example.apigatewayservice.filter;

import lombok.Data;
import lombok.extern.slf4j.Slf4j;
import org.springframework.cloud.gateway.filter.GatewayFilter;
import org.springframework.cloud.gateway.filter.OrderedGatewayFilter;
import org.springframework.cloud.gateway.filter.factory.AbstractGatewayFilterFactory;
import org.springframework.core.Ordered;
import org.springframework.http.HttpHeaders;
import org.springframework.http.HttpStatus;
import org.springframework.http.server.reactive.ServerHttpResponse;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
@Slf4j
public class CorsFilter extends AbstractGatewayFilterFactory<CorsFilter.Config> {

    public CorsFilter(){super(Config.class);}

    @Override
    public GatewayFilter apply(CorsFilter.Config config) {
        return new OrderedGatewayFilter((exchange, chain) -> {
            log.info("CORS filter applied for request: {}", exchange.getRequest().getURI());
            log.info("Request method: {}", exchange.getRequest().getMethodValue());

            ServerHttpResponse response = exchange.getResponse();
            HttpHeaders headers = response.getHeaders();
            headers.add("Access-Control-Allow-Origin", "*");
            headers.add("Access-Control-Allow-Methods", "POST, GET, DELETE, PUT, OPTIONS");
            headers.add("Access-Control-Allow-Headers", "Origin, X-Requested-With, Content-Type, Accept, Authorization, Set-Cookie, X-Auth-Token");
            headers.add("Access-Control-Max-Age", "3600");

            if ("OPTIONS".equalsIgnoreCase(exchange.getRequest().getMethodValue())) {
                response.setStatusCode(HttpStatus.OK);
                return Mono.empty();
            } else {
                return chain.filter(exchange);
            }
        }, Ordered.HIGHEST_PRECEDENCE); // 여기서 가장 높은 우선순위를 설정
    }

    @Data
    public static class Config{
        private String baseMessage;
        private boolean preLogger;
        private boolean postLogger;

    }
}

File: src/main/kotlin/com/miniproject/kubeBoard/config/CorsFilter.kt
package com.miniproject.kubeBoard.config

import org.springframework.stereotype.Component
import javax.servlet.*
import javax.servlet.FilterConfig
import javax.servlet.ServletRequest
import javax.servlet.http.HttpServletRequest
import javax.servlet.http.HttpServletResponse

@Component
class CorsFilter : Filter {

    override fun init(filterConfig: FilterConfig) {
        // Nothing to initialize
    }

    override fun doFilter(req: ServletRequest, res: ServletResponse, chain: FilterChain) {
        val request = req as HttpServletRequest
        val response = res as HttpServletResponse

        response.apply {
            setHeader("Access-Control-Allow-Origin", "*")
            setHeader("Access-Control-Allow-Credentials", "true")
            setHeader("Access-Control-Allow-Methods", "POST, GET, DELETE, PUT, OPTIONS")
            setHeader(
                    "Access-Control-Allow-Headers",
                    "Origin, X-Requested-With, Content-Type, Accept, Authorization, Set-Cookie, X-Auth-Token"
            )
            setHeader("Access-Control-Max-Age", "3600")
        }
        //요청의 HTTP 메서드가 OPTIONS인 경우 (CORS의 preflight 요청), 응답 상태를 200 OK로 설정하고 처리를 종료
        if ("OPTIONS".equals(request.method, ignoreCase = true)) {
            response.status = HttpServletResponse.SC_OK
        } else {
            //필터 체인의 다음 단계 진행
            chain.doFilter(req, res)
        }
    }

    override fun destroy() {
        // Nothing to destroy
    }
}